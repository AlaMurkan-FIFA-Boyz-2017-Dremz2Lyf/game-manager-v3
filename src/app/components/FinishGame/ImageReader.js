import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { Segment, Progress, Message } from 'semantic-ui-react';

import Tesseract from 'tesseract.js';

const whitelist = 'AYRSTPOTFIC%0123456789hotsnfargjetiwckluyd ';

class ImageReader extends PureComponent {
  static propTypes = {
    images: PropTypes.array,
  }
  state = {
    statsPercent: 0,
    scorePercent: 0,
    results: null,
  }

  tesseractJob = Tesseract.create({
    workerPath: 'https://cdn.rawgit.com/naptha/tesseract.js/1.0.10/dist/worker.js',
    langPath: 'https://cdn.rawgit.com/naptha/tessdata/gh-pages/3.02/',
  });

  errorMessage = 'Welp, something went wrong. Tell Scott to fix his crap.'

  readImages = (images) => {
    const promises = images.map((image) => {
      console.log({ image });
      return Promise.resolve(this.tesseractJob.recognize(image, {
        lang: 'eng',
        tessedit_char_whitelist: whitelist,
      }).progress(progress => console.log({ progress })).catch((error) => {
        this.setState({ error });
      }));
    });

    return Promise.all(promises).then((results) => {
      const final = results.map(({ lines }) => lines.map(({ text }) => text));
      this.tesseractJob.terminate();
      return final;
    }).then((results) => {
      this.setState({
        results,
      });
    });
  }

  render() {
    const { statsPercent, scorePercent, results, error } = this.state;
    const { images } = this.props;
    if (images.length === 2 && !results) {
      this.readImages(images);
    }

    return (
      <Segment basic>
        <Message color="teal">
          <Message.Header>
            Processing Image
          </Message.Header>
        </Message>
        {
          error ? (
            <Message color="red" header="Oops!" content={this.errorMessage} />
          ) : (
            <Segment basic>
              <p>{results}</p>
              <Progress percent={scorePercent}>
                Processing Score
              </Progress>
              <Progress percent={statsPercent}>
                Processing Stats
              </Progress>
            </Segment>
          )
        }
      </Segment>
    );
  }
}

export default ImageReader;
