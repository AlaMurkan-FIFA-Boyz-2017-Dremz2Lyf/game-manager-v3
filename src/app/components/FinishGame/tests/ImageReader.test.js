import React from 'react';
import { shallow } from 'enzyme';

import ImageReader from '../ImageReader';
import TesseractMock from './tesseractMock';

const images = { score: 'https://goo.gl/images/wCqoJT', stats: 'https://goo.gl/images/wCqoJT' };

const tessMock = new TesseractMock({ shouldError: false });

describe('ImageReader Component', () => {
  let wrapper;
  let readImagesSpy;
  let onResults;

  beforeEach(() => {
    readImagesSpy = jest.spyOn(ImageReader.prototype, 'readImages');
    onResults = jest.fn();
  });

  afterEach(() => {
    onResults.mockReset();
    readImagesSpy.mockReset();
  });

  it('should match snapshot and show the progress tracker', () => {
    wrapper = shallow(
      <ImageReader processer={tessMock} images={images} onResults={onResults} />,
    );
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should match the snapshot if there are no images passed to the component', () => {
    wrapper = shallow(
      <ImageReader processer={tessMock} images={{}} onResults={onResults} />,
    );
    expect(wrapper.html()).toMatchSnapshot();
    expect(wrapper.find('Message').length).toBe(2);
    expect(wrapper.find('Message').last().prop('color')).toBe('red');
  });

  it('should call readImages if there are images passed to the component', () => {
    wrapper = shallow(<ImageReader processer={tessMock} images={images} onResults={onResults} />);

    expect(readImagesSpy).toHaveBeenCalled();
  });

  it('should not call readImages if there are no images', () => {
    wrapper = shallow(<ImageReader processer={tessMock} />);
    expect(readImagesSpy).not.toHaveBeenCalled();
  });

  describe('readImages', () => {
    it('should call onResults', () => {
      readImagesSpy.mockRestore();
      wrapper = shallow(<ImageReader processer={tessMock} images={images} onResults={onResults} />);

      return wrapper.instance().readImages(images)
        .then(() => expect(onResults).toHaveBeenCalled());
    });
  });
});
