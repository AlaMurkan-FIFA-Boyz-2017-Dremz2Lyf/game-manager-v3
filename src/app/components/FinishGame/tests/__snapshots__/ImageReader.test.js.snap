// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImageReader Component should match snapshot and show the progress tracker 1`] = `"<div class=\\"ui basic segment\\"><div class=\\"ui teal message\\"><div class=\\"header\\">Processing Image</div></div><div class=\\"ui basic segment\\"><div class=\\"ui teal progress\\" data-percent=\\"100\\"><div class=\\"bar\\" style=\\"width:100%\\"><div class=\\"progress\\">100%</div></div><div class=\\"label\\"><div class=\\"ui grey small header\\">Processing Score</div></div></div><div class=\\"ui teal progress\\" data-percent=\\"100\\"><div class=\\"bar\\" style=\\"width:100%\\"><div class=\\"progress\\">100%</div></div><div class=\\"label\\"><div class=\\"ui grey small header\\">Processing Stats</div></div></div></div></div>"`;

exports[`ImageReader Component should match the snapshot if there are no images passed to the component 1`] = `"<div class=\\"ui basic segment\\"><div class=\\"ui teal message\\"><div class=\\"header\\">Processing Image</div></div><div class=\\"ui red message\\"><div class=\\"content\\"><div class=\\"header\\">Oops!</div><p>Welp, something went wrong. Tell Scott to fix his crap.</p></div></div></div>"`;
